### PR TITLE
Prevent websocket upgrade from replying with `content-length: 0`

### PR DESCRIPTION
**Pull request by chonton**
Fixes [quarkusio/quarkus #370505](https://github.com/quarkusio/quarkus/issues/37050)

Additionally, using the native vertx implementation increases the number of websocket clients and versions supported.

### CODE REVIEW

File: websocket/vertx/src/main/java/io/undertow/websockets/vertx/VertxWebSocketHttpExchange.java [hunk 1]
```diff
@@ -118,17 +118,10 @@ public void upgradeChannel(Consumer<Object> listener) {
...
         }
 
-        response.setStatusCode(101).end(new Handler<AsyncResult<Void>>() {
-            @Override
-            public void handle(AsyncResult<Void> event) {
-                Http1xServerConnection connection = (Http1xServerConnection) request.connection();
-                ChannelHandlerContext context = connection.channelHandlerContext();
-                context.pipeline().remove("httpDecoder");
-                context.pipeline().remove("httpEncoder");
+        request.toWebSocket().onSuccess( ws -> {
```
Comment (cescoffier) [inline websocket/vertx/src/main/java/io/undertow/websockets/vertx/VertxWebSocketHttpExchange.java:121]: Avoid lambda please.

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Comment by cescoffier** [comment]
Would be great to have a test case.

**Comment by chonton** [comment]
test case is in quarkus integration tests. please note that there currently aren't any tests for the vertx code

File: websocket/vertx/src/main/java/io/undertow/websockets/vertx/VertxWebSocketHttpExchange.java (hunk 1 above)
Comment (chonton) [reply websocket/vertx/src/main/java/io/undertow/websockets/vertx/VertxWebSocketHttpExchange.java:121]: done

**Comment by cescoffier** [comment]
Ah yes... I forgot that this project still use junit 4.

**Review by cescoffier — APPROVED** [review]

**Review by stuartwdouglas — APPROVED** [review]

**Comment by cescoffier** [comment]
@chonton I will revert this PR, unfortunately.  
It is breaking sub-protocols (see https://github.com/quarkusio/quarkus/issues/39082). I'm not totally sure why. Let's revert and investigate.

**Comment by michaelfrid** [comment]
@chonton @cescoffier 
Hello. We just encounter this issue in Quarkus 3,2,X and 3.15.X. It causes our .Net clients to close websocket connection after some period of time. Are there any plans to fix it ?
Thanks a lot

**Comment by cescoffier** [comment]
Did you try with WebSocket Next?

**Comment by cescoffier** [comment]
(We reverted the fix as it introduced a regression)

**Comment by michaelfrid** [comment]
I see it's marked as experimental feature. Is it stable enough to be used in the production environment ?
It also means we need to rewrite our current implementation, right ?

**Comment by cescoffier** [comment]
Yes, it is not quite mature. It is going to become our recommended approach
for WebSockets.

Le jeu. 9 janv. 2025 à 11:31, michaelfrid ***@***.***> a
écrit :

**Comment by michaelfrid** [comment]
@cescoffier
So what should we do for meanwhile ? Any chances the original issue will be fixed ?

**Comment by cescoffier** [comment]
Sorry, it is quite mature :-) 
I would recommend switching.

I don't believe this issue will be fixed anytime soon. At least I do not have the bandwidth.

**Pull request merged**